src/CachePeer.cc

```diff
@@ -131,3 +136,421 @@ operator <<(std::ostream &os, const CachePeer &p)
+            p->rename(token + 5);
+        } else if (!strncmp(token, "forceddomain=", 13)) {
+            safe_free(p->domain);
+            if (token[13])
+                p->domain = xstrdup(token + 13);
+
+        } else if (strncmp(token, "ssl", 3) == 0) {
+#if !USE_OPENSSL
+            debugs(0, DBG_CRITICAL, "WARNING: cache_peer option '" << token << "' requires --with-openssl");
+#else
+            p->secure.parse(token+3);
+#endif
+        } else if (strncmp(token, "tls-", 4) == 0) {
+            p->secure.parse(token+4);
+        } else if (strncmp(token, "tls", 3) == 0) {
+            p->secure.parse(token+3);
+        } else if (strcmp(token, "front-end-https") == 0) {
+            p->front_end_https = 1;
+        } else if (strcmp(token, "front-end-https=on") == 0) {
+            p->front_end_https = 1;
+        } else if (strcmp(token, "front-end-https=auto") == 0) {
+            p->front_end_https = 2;
+        } else if (strcmp(token, "connection-auth=off") == 0) {
+            p->connection_auth = 0;
+        } else if (strcmp(token, "connection-auth") == 0) {
+            p->connection_auth = 1;
+        } else if (strcmp(token, "connection-auth=on") == 0) {
+            p->connection_auth = 1;
+        } else if (strcmp(token, "connection-auth=auto") == 0) {
+            p->connection_auth = 2;
+        } else if (token[0] == '#') {
+            // start of a text comment. stop reading this line.
+            break;
+        } else {
+            debugs(3, DBG_PARSE_NOTE(DBG_IMPORTANT), "ERROR: Ignoring unknown cache_peer option '" << token << "'");
+        }
+    }
+
+    if (findCachePeerByName(p->name))
+        throw TextException(ToSBuf("cache_peer ", *p, " specified twice"), Here());
+
+    if (p->max_conn > 0 && p->max_conn < p->standby.limit)
+        throw TextException(ToSBuf("cache_peer ", *p, " max-conn=", p->max_conn,
+                                   " is lower than its standby=", p->standby.limit), Here());
+
+    if (p->weight < 1)
+        p->weight = 1;
+
+    if (p->connect_fail_limit < 1)
+        p->connect_fail_limit = 10;
+
+#if USE_CACHE_DIGESTS
+    if (!p->options.no_digest)
+        p->digest = new PeerDigest(p);
+#endif
+
+    if (p->secure.encryptTransport)
+        p->secure.parseOptions();
+
+    cachePeers.emplace_back(p);
+
+    p->index = cachePeers.size();
+
+    peerClearRRStart();
+}
+
+CachePeers::CachePeerList::const_iterator
+CachePeers::firstPing() const
+{
+    auto it = cachePeers.begin();
+    std::advance(it, firstPing_);
+    return it;
+}
+
+void
+CachePeers::advanceFirstPing()
+{
+    assert(firstPing_ <= size());
+    if (++firstPing_== size())
+        firstPing_ = 0;
+}
+
+void
+CachePeers::remove(CachePeer *p)
+{
+    for (auto it = cachePeers.begin(); it != cachePeers.end(); ++it) {
+        if (it->get() == p) {
+            cachePeers.erase(it);
+            break;
+        }
+    }
+    firstPing_ = 0;
+}
+
+CachePeers &
```

Comment:
A caller that removed CachePeer needed a non-constant reference. I adjusted it, creating a dedicated function and then made this function return const reference.

Also, since first I made our begin()/end() methods non-const (as standard ones), this did not work with for-range loops (for-range loops do not use cbegin()/cend()). So to fix those loops I had to make begin()/end() const and removed (unused) cbegin()/cend().